# Beautiful Tables

**Inhalt:** Lesbare Tabellen in Pandas

**Nötige Skills:** Erste Schritte mit Pandas

**Lernziele:**
- Zahlenformate kennenlernen und anwenden
- Tabellen-Leserlichkeit mit Styles optimieren
- Chartform: Heatmaps

**Weitere Ressourcen:**
- Cookbook für String Formats: https://mkaz.blog/code/python-string-format-cookbook/
- User Guide zu Styles: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.io.formats.style.Styler.format.html
- Reference Style-Funktion: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.io.formats.style.Styler.format.html

In [3]:
import pandas as pd

In [33]:
import numpy as np

## Das Beispiel

Eine Liste von Ländern mit ihrer Grösse und diversen Eigenschaften.

Quelle: Weltbank (https://data.worldbank.org/indicator)

## Daten laden

In [4]:
path = "dataprojects/Worldbank/worldbank_countries.xlsx"

In [5]:
df = pd.read_excel(path)

In [6]:
df.head(3)

Country Name Country Code  Population  Forest Area  GDP per Capita  \
0                Kuwait          KWT     4270563     0.003507           36290   
1  United Arab Emirates          ARE     9890400     0.044678           43470   
2     Brunei Darussalam          BRN      437483     0.721063           32230   

   Unemployment  Renewable Energy  Life Expectancy  \
0        0.0679          0.000336           75.489   
1        0.0500          0.002592           77.972   
2        0.0837          0.000109           75.860   

   Female Labor Participation  Urban Population  CO2 Emissions per Capita  \
0                      0.4978           1.00000                 21.622724   
1                      0.5205           0.87048                 20.797498   
2                      0.5609           0.78250                 16.644909   

   Fertility Rate  Population Growth  
0           2.077           0.014978  
1           1.387           0.012194  
2           1.819           0.009617

## Pandas Display Options

Wir interessieren uns also für Länder und deren diverse Eigenschaften. Was in Pandas zuweilen aber gar nicht so easy ist: Einfach mal durch die Daten schauen.

Dataframes werden in der Übersicht abgekürzt, wir sehen nur die ersten und letzten fünf Zeilen:

In [9]:
df

Country Name Country Code  Population  Forest Area  \
0                      Kuwait          KWT     4270563     0.003507   
1        United Arab Emirates          ARE     9890400     0.044678   
2           Brunei Darussalam          BRN      437483     0.721063   
3                      Canada          CAN    38005238     0.386955   
4                   Australia          AUS    25687041     0.174213   
..                        ...          ...         ...          ...   
150  Central African Republic          CAF     4829764     0.358005   
151                      Chad          TCD    16425859     0.034252   
152                   Burundi          BDI    11890781     0.108894   
153                   Somalia          SOM    15893219     0.095323   
154          Congo, Dem. Rep.          COD    89561404     0.556473   

     GDP per Capita  Unemployment  Renewable Energy  Life Expectancy  \
0             36290        0.0679          0.000336         75.48900   
1             43470        0.0500          0.002592         77.97200   
2             32230        0.0837          0.000109         75.86000   
3             46460        0.0948          0.221829         82.04878   
4             55100        0.0661          0.096450         82.90000   
..              ...           ...               ...              ...   
150             520        0.0433          0.809485         53.28300   
151             700        0.0226          0.852578         54.23900   
152             280        0.0080          0.855269         61.58400   
153             320        0.1310          0.948753         57.39700   
154             530        0.0455          0.963837         60.68100   

     Female Labor Participation  Urban Population  CO2 Emissions per Capita  \
0                        0.4978           1.00000                 21.622724   
1                        0.5205           0.87048                 20.797498   
2                        0.5609           0.78250                 16.644909   
3                        0.6113           0.81562                 15.497025   
4                        0.6082           0.86241                 15.475516   
..                          ...               ...                       ...   
150                      0.6447           0.42198                  0.070719   
151                      0.6400           0.23520                  0.069132   
152                      0.8050           0.13708                  0.052795   
153                      0.2221           0.46141                  0.045975   
154                      0.6099           0.45638                  0.026169   

     Fertility Rate  Population Growth  
0            2.0770           0.014978  
1            1.3870           0.012194  
2            1.8190           0.009617  
3            1.4684           0.010896  
4            1.6570           0.012587  
..              ...                ...  
150          4.6450           0.017668  
151          5.6490           0.029594  
152          5.3210           0.030761  
153          5.9780           0.028743  
154          5.8190           0.031427  

[155 rows x 13 columns]

Ein erster, simpler Schritt: Wir wollen einfach mal sämtliche Daten sehen, statt nur die Vorschau der ersten und letzten fünf Zeilen. Wir können dazu eine Einstellung vornehmen, die ab diesem Punkt für das gesamte weitere Notebook gilt:

In [10]:
pd.set_option("display.max_rows", 1000)

Ab jetzt zeigt Pandas jedesmal bis zu 1000 Zeilen einer Tabelle an. Das genügt uns.

In [11]:
df

Country Name Country Code  Population  Forest Area  \
0                            Kuwait          KWT     4270563     0.003507   
1              United Arab Emirates          ARE     9890400     0.044678   
2                 Brunei Darussalam          BRN      437483     0.721063   
3                            Canada          CAN    38005238     0.386955   
4                         Australia          AUS    25687041     0.174213   
5                      Saudi Arabia          SAU    34813867     0.004545   
6                     United States          USA   329484123     0.338669   
7                              Oman          OMN     5106622     0.000081   
8               Trinidad and Tobago          TTO     1399491     0.444815   
9                      Turkmenistan          TKM     6031187     0.087822   
10                      Korea, Rep.          KOR    51780579     0.644688   
11                          Estonia          EST     1331057     0.560939   
12               Russian Federation          RUS   144104080     0.497843   
13                   Czech Republic          CZE    10698896     0.346773   
14                            Libya          LBY     6871287     0.001233   
15                      Netherlands          NLD    17441139     0.109742   
16                            Japan          JPN   125836021     0.684088   
17                          Germany          DEU    83240525     0.326836   
18                        Singapore          SGP     5685807     0.219605   
19                           Poland          POL    37950802     0.309730   
20                          Belgium          BEL    11555997     0.227642   
21                          Ireland          IRL     4994724     0.113517   
22                         Malaysia          MYS    32365998     0.581770   
23                     South Africa          ZAF    59308690     0.140551   
24                            China          CHN  1402112000     0.233406   
25                          Austria          AUT     8917205     0.472510   
26                           Norway          NOR     5379475     0.333600   
27           Bosnia and Herzegovina          BIH     3280815     0.427326   
28                         Slovenia          SVN     2100126     0.614723   
29                         Mongolia          MNG     3278292     0.091011   
30                      New Zealand          NZL     5084300     0.375701   
31                           Serbia          SRB     6908224     0.311302   
32                          Belarus          BLR     9398861     0.431944   
33                           Greece          GRC    10715549     0.302700   
34                           Cyprus          CYP     1207361     0.186721   
35                  Slovak Republic          SVK     5458827     0.400562   
36                     Bahamas, The          BHS      393248     0.509351   
37                         Bulgaria          BGR     6927288     0.358604   
38                          Denmark          DNK     5831404     0.157110   
39                            Spain          ESP    47351567     0.371738   
40                   United Kingdom          GBR    67215293     0.131856   
41                            Italy          ITA    59554023     0.321302   
42                Equatorial Guinea          GNQ     1402985     0.872877   
43                           Turkey          TUR    84339067     0.288715   
44                         Portugal          PRT    10305564     0.361550   
45                          Hungary          HUN     9749763     0.224963   
46                            Chile          CHL    19116209     0.244922   
47                           France          FRA    67391582     0.315090   
48                      Switzerland          CHE     8636896     0.321163   
49                         Barbados          BRB      287371     0.146512   
50                          Ukraine          UKR    44134693     0.167242   
51                 

Für eine unbeschränkte Ansicht könnten wir hier statt einem Zahlenwert auch `None` setzen.

Es gibt weitere Display-Optionen, die wir einstellen können:
- `display.max_columns`: Maximale Spaltenzahl, analog zur Zeilenzahl. Damit wir auch alle Spalten sehen, falls es sehr viele sind.
- `display.max_colwidth`: Maximale Zeichenzahl, die in einer Zelle angezeigt werden soll. Damit der ganze Text in einer Zelle angezeigt wird, falls wir nicht nur eine abgekürzte Variante davon sehen wollen.

Eine vollständige Liste der Optionen findet sich hier: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

Praktisch für uns ist zum Beispiel auch die folgende Option:

In [12]:
pd.set_option("display.precision", 2)

Wir erhalten nun nicht mehr zig Dezimalstellen angezeigt, sondern nur noch eine. Das erleichtert das Lesen.

In [13]:
df.head(5)

Country Name Country Code  Population  Forest Area  GDP per Capita  \
0                Kuwait          KWT     4270563     3.51e-03           36290   
1  United Arab Emirates          ARE     9890400     4.47e-02           43470   
2     Brunei Darussalam          BRN      437483     7.21e-01           32230   
3                Canada          CAN    38005238     3.87e-01           46460   
4             Australia          AUS    25687041     1.74e-01           55100   

   Unemployment  Renewable Energy  Life Expectancy  \
0          0.07          3.36e-04            75.49   
1          0.05          2.59e-03            77.97   
2          0.08          1.09e-04            75.86   
3          0.09          2.22e-01            82.05   
4          0.07          9.65e-02            82.90   

   Female Labor Participation  Urban Population  CO2 Emissions per Capita  \
0                        0.50              1.00                     21.62   
1                        0.52              0.87                     20.80   
2                        0.56              0.78                     16.64   
3                        0.61              0.82                     15.50   
4                        0.61              0.86                     15.48   

   Fertility Rate  Population Growth  
0            2.08           1.50e-02  
1            1.39           1.22e-02  
2            1.82           9.62e-03  
3            1.47           1.09e-02  
4            1.66           1.26e-02

Das ist besser, aber immer noch nicht schön. Zum Beispiel würden wir die Arbeitslosenquote gerne in Prozent sehen (7,1% statt 0.07) und hätten bei der Bevölkerung gerne Tausender-Trennzeichen (4'270'563).

Um das zu tun, bietet Pandas / Python eine Reihe von Formatierungscodes an.

## Exkurs: Zahlen formatieren

### Das Prinzip

Zahlen zu formatieren, funktioniert in Python etwas schräg.

Wir starten mit einem Format-Ausdruck (einem String):

In [16]:
'{:.2f}'

'{:.2f}'

Ausgehend von diesem String können wir eine Funktion aufrufen: `.format()`

Der Parameter, den wir dieser Funktion mitgeben, ist die Zahl, die wir formatieren wollen.

In [17]:
'{:.2f}'.format(5.39120)

'5.39'

Note: Bei dieser Prozedur haben wir auch eine Typenumwandlung vollzogen:
- Unsere ursprüngliche Zahl (`5.39120`) war eine Floating-Point-Number

In [18]:
type(5.39120)

float

- Der Output dieser Funktion (`'5.39'`) ist ein String.

In [19]:
type('5.39')

str

Nach dem Formatieren können wir also keine mathematischen Berechnungen mehr mit der Zahl machen!

### Format-Strings

Wie funktionieren diese Formatierungs-Strings?

#### 1. Die Hülle

Der einfachste Formatstring macht - gar nichts, sondern gibt die Zahl einfach tel quel zurück: `{}`

In [20]:
'{}'.format(3.1415926)

'3.1415926'

Damit wir die Zahl bearbeiten können, braucht es zunächst mal einen Doppelpunkt: `:`

Wenn wir den einsetzen, passiert aber immer noch nichts.

In [21]:
'{:}'.format(3.1415926)

'3.1415926'

Es braucht also noch weitere Angaben.

#### 2. Runden

Wir können eine Zahl nach dem Komma runden mit folgendem Ausdruck: `.2f`

Das f steht dabei für "float" und die 2 für die Anzahl Ziffern nach dem Komma.

In [22]:
'{:.2f}'.format(3.1415926)

'3.14'

#### 3. Prozentzahlen

Um die Zahl in Prozent anzugeben, Prozentzeichen: `%`

In [23]:
'{:%}'.format(3.1415926)

'314.159260%'

Man kann auch Prozentzahlen runden:

In [24]:
'{:.2%}'.format(3.1415926)

'314.16%'

#### 4. Plus- und Minuszeichen

Um explizit ein Pluszeichen anzugeben, falls die Zahl grösser als null ist: `+`

In [25]:
'{:+.2%}'.format(3.1415926)

'+314.16%'

#### 5. Nullen vorneanstellen

Kann zB hilfreich sein, wenn man die Zahlen später als Text sortieren muss oder wenn es sich um bestimmte Codes handelt: `03d`

Das "d" steht dabei für digits (ganze Zahlen) und die 3 für die Gesamtzahl der Ziffern.

In [15]:
'{:03d}'.format(42)

'042'

Das funktioniert ähnlich auch bei Floats. Allerdings muss man hier sowohl die Gesamtzahl der Zeichen (zB 6) als auch die Länge hinter dem Komma angeben (zB 2):

In [16]:
'{:06.2f}'.format(3.1415926)

'003.14'

**6. Tausendertrennzeichen**

Trägt auch stark zur Lesbarkeit bei: Tausenderzeichen. Der Code dafür ist ein Komma: `,`

In [19]:
'{:,}'.format(31415926)

'31,415,926'

Leider müssen wir uns mit der amerikanischen Variante (,) begnügen. Die deutschen Hochkommas (') gehen nicht.

#### 7. Währungen

Um eine bestimmte Währung vor oder hinter der Zahl anzuzeigen: Einfach ausserhalb der geschweiften Klammern die Einheit hinschreiben, zB `CHF`

In [20]:
'CHF {:,}'.format(31415926)

'CHF 31,415,926'

#### 8. Grosse zahlen runden

Wenn zB die genauen Frankenbeträge nicht wichtig sind, können wir sie auch vor dem Komma runden.

In [30]:
# code fehlt noch. gibt es das überhaupt?

**Übersicht und weitere Formate:**

Hier eine Übersicht, kopiert von: https://mkaz.blog/code/python-string-format-cookbook/

| Number | Format | Output | Description |
| -------|--------|--------|------------- |
| 3.1415926 | {:.2f} | 3.14 | Format float 2 decimal places |
| 3.1415926 | {:+.2f} | +3.14 | Format float 2 decimal places with sign |
| -1 | {:+.2f} | -1.00 | Format float 2 decimal places with sign |
| 2.71828 | {:.0f} | 3 | Format float with no decimal places |
| 5 | {:0>2d} | 05 | Pad number with zeros (left padding, width 2) |
| 5 | {:x<4d} | 5xxx | Pad number with x’s (right padding, width 4) |
| 10 | {:x<4d} | 10xx | Pad number with x’s (right padding, width 4) |
| 1000000 | {:,} | 1,000,000 | Number format with comma separator |
| 0.25 | {:.2%} | 25.00% | Format percentage |
| 1000000000 | {:.2e} | 1.00e+09 | Exponent notation |
| 13 | {:10d} |         13 | Right aligned (default, width 10) |
| 13 | {:<10d} | 13 | Left aligned (width 10) |
| 13 | {:^10d} |     13 | Center aligned (width 10) |

Weitere Infos zB hier: https://pyformat.info/

## Tabellen separat formatieren

Wir können die Format-Strings nun nutzen, um Spalten separat zu formatieren. 

Dazu definieren wir in einem Dictionary für jede Spalte das passende Format.

In [23]:
col_formats = {
    'Country Name': '{}',
    'Country Code': '{}',
    'Population': '{:,}',
    'Forest Area': '{:.0%}',
    'GDP per Capita': '$ {:,}',
    'Unemployment': '{:.0%}',
    'Renewable Energy': '{:.0%}',
    'Life Expectancy': '{:.1f}',
    'Female Labor Participation': '{:.0%}',
    'Urban Population': '{:.0%}',
    'CO2 Emissions per Capita': '{:.1f} t',
    'Fertility Rate': '{:.1f}',
    'Population Growth': '{:.1%}'
}

Wir können diesen Dictionary verwenden, um den `style` der Tabelle zu definieren:

In [65]:
df.style.format(col_formats)

**Achtung:** Was wir hier sehen, ist kein Dataframe mehr, sondern nur noch eine Repräsentation davon.

Der folgende Code funktioniert deshalb nicht:

In [25]:
# df.style.format(col_formats).head()

AttributeError: 'Styler' object has no attribute 'head'

Das heisst leider: Wir können den Style eines Dataframes immer nur ganz am Schluss setzen, wenn wir keine Berechnungen mehr machen müssen und nur noch das Endergebnis sehen wollen.

Aber immerhin: Wir können die definierten Spaltenformate auf beliebige Modifikationen unseres Dataframes anwenden, sofern einzelne Spalten immmer noch gleich heissen.

In [31]:
df[df['Country Name'] == 'Switzerland'][['Population', 'GDP per Capita']].style.format(col_formats)

## Coole Dinge, die man mit Styles tun kann

Da wir nun aber schonmal mit Styles arbeiten, hier einige hübsche Anwendungen, die es uns noch leichter machen, die Daten in unserem DF zu lesen, Muster zu erkennen und evtl auch Fehler zu bemerken.

### Einzelne Werte herausstreichen

#### 1. Fehlende Zahlen

Angenommen, es hätte sich irgendwo ein `NaN`-Wert in unsere Daten reingeschlichen:

In [34]:
df.loc[df['Country Name'] == 'Bulgaria', 'Renewable Energy'] = np.nan

Es gibt eine einfache Style-Funktion, die uns darauf hinweist: `highlight_null()`

In [35]:
df.style.highlight_null(null_color='yellow')

#### Kleinste und grösste Werte

Wir können für jede Spalte anzeigen:
- wo der Maximalwert zu finden ist: `highlight_max()`
- wo der Minimalwert zu finden ist: `highlight_min()`

In [37]:
df.style.highlight_max()

### Farbcodierte Werte

Wir können Styles auch benutzen, um eine Tabelle Heatmap-ähnlich einzufärben.

Dazu gibt es die Funktion `background_gradient()`:

In [44]:
df[['Country Name', 'Unemployment']].style.background_gradient(cmap='coolwarm')

Uns stehen eine Reihe von verschiedenen Colormaps zur Verfügung: https://matplotlib.org/2.0.2/examples/color/colormaps_reference.html

Zusätzlich können wir definieren, an welchen Eckpunkten die Range angesetzt werden soll.

Zum Beispiel:
- Minimum bei 0.05 (unterhalb davon ist alles blau)
- Maximum bei 0.15 (oberhalb davon ist alles tiefrot)

In [45]:
df[['Country Name', 'Unemployment']].style.background_gradient(cmap='coolwarm', vmin=0.05, vmax=0.15)

Wir können Colormaps auch nur auf einzelne Spalten anwenden mit dem Parameter `subset=`

In [46]:
df.style.background_gradient(cmap='coolwarm', vmin=0.05, vmax=0.15, subset=['Unemployment'])

Und wir können unterschiedliche Colormaps auf unterschiedliche Spalten anwenden.

Dazu verketten wir einfach zwei `background_gradient()` aneinander:

In [49]:
df.style.background_gradient(
    cmap='coolwarm',
    vmin=0.05,
    vmax=0.15,
    subset=['Unemployment']
).background_gradient(
    cmap='Greens',
    vmin=0,
    vmax=1,
    subset=['Forest Area']
)

Diese Verkettung funktioniert auch, wenn wir zum Schluss noch die Zahlen formatieren wollen:

In [50]:
df.style.background_gradient(
    cmap='coolwarm',
    vmin=0.05,
    vmax=0.15,
    subset=['Unemployment']
).background_gradient(
    cmap='Greens',
    vmin=0,
    vmax=1,
    subset=['Forest Area']
).format(col_formats)

### Mini Bar Charts

Eine weitere Darstellungsmöglichkeit: kleine Barcharts in die Tabelle integrieren mit `bar()`:

In [53]:
df.style.bar(subset=['Urban Population'], color='#d65f5f')

## Styles wiederverwenden

Sagen wir, wir haben einen wunderschönen Style generiert – inklusive Heatmap, Barchart, Highlights und Zahlenformaten.

In [60]:
df.style.background_gradient(
    cmap='coolwarm',
    vmin=0.05,
    vmax=0.15,
    subset=['Unemployment']
).background_gradient(
    cmap='Greens',
    vmin=0,
    vmax=1,
    subset=['Forest Area']
).bar(
    subset=['Urban Population'],
    color='#d65f5f'
).format(
    col_formats
).highlight_max().highlight_min()

Um diesen Style wiederzuverwenden, müssen wir nicht jedesmal den ganzen Wust erneut schreiben.

Wir können den Style speichern mit `export()`...

In [62]:
my_style = df.style.background_gradient(
    cmap='coolwarm',
    vmin=0.05,
    vmax=0.15,
    subset=['Unemployment']
).background_gradient(
    cmap='Greens',
    vmin=0,
    vmax=1,
    subset=['Forest Area']
).bar(
    subset=['Urban Population'],
    color='#d65f5f'
).format(
    col_formats
).highlight_max().highlight_min().export()

... und den Style erneut anwenden mit `use()`:

In [63]:
df.style.use(my_style)

## Formatierte Daten exportieren

### nach Excel

Wir können formatierte Dataframes ins Excel transportieren. Das funktioniert aber nur mässig: Bei unserem Beispiel klappt es mit den Heatmaps und Highlights, aber nicht mit den Barcharts und Zahlenformaten :-/

In [73]:
df.style.use(my_style).to_excel("dataprojects/Worldbank/formatted_table.xlsx", index=False)

### als csv

Beim Export in ein csv sollte man sich grundsätzlich gut überlegen, wie viel Formatierung man überhaupt exportieren will!
- **Tausenderzeichen** im CSV festzuhalten macht sehr wahrscheinlich keinen Sinn, da das nächste Programm, das diese Daten bearbeitet, diese erst wieder mühsam erkennen will.
- Dasselbe gilt für **Währungsangaben** oder auch für **Prozentzeichen**. Normalerweise würde man diese Formatierungen im weiterverarbeitenden Programm (zB in einem Charttool) wieder separat vornehmen. Das heisst, wir exportieren nicht "5.1%" sondern "0.051".
- Was allerdings Sinn machen kann, ist, Zahlen zu runden, also die Zahl der **Dezimalstellen** zu beschränken. Dafür gibt es in der Export-Funktion einen passenden Parameter.

In [74]:
df.to_csv("dataprojects/Worldbank/rounded_table.csv", float_format="%.2f")